torchvision/prototype/datasets/_builtin/cityscapes.py

```diff
@@ -0,0 +1,201 @@
+            filenames = (f"leftImg8bit_{split_label}.zip", "gtCoarse.zip")
+
+        resources: List[OnlineResource] = [
+            CityscapesResource(file_name=file_name, sha256=self._FILES_CHECKSUMS[file_name]) for file_name in filenames
+        ]
+        return resources
+
+    def _filter_split_images(self, data: Tuple[str, Any], *, req_split: str) -> bool:
+        path = Path(data[0])
+        split = path.parent.parts[-2]
+        return split == req_split and ".png" == path.suffix
+
+    def _filter_group_targets(self, data: Tuple[str, Any]) -> str:
+        path = Path(data[0])
+        # Taregt path looks like
+        # gtFine/val/frankfurt/frankfurt_000001_066574_gtFine_polygons.json
+        # and we produce the key: frankfurt_000001_066574
+        return path.name.rsplit("_", 1)[0]
+
+    def _prepare_sample(self, data: Tuple[Tuple[str, Any], Any]) -> Dict[str, Any]:
+        (img_path, img_data), target_data = data
+
+        output = dict(image_path=img_path, image=EncodedImage.from_file(img_data))
+        # reorder inside group of targets and setup output dictionary:
+        for path, data in target_data:
+            stem = Path(path).stem
+            for target_type in ["instance", "label", "polygon", "color"]:
+                if target_type in stem:
+                    if target_type == "polygon":
+                        enc_data = json.loads(data.read())
+                    else:
+                        enc_data = EncodedImage.from_file(data)
+                    output[target_type] = enc_data
+                    output[f"{target_type}_path"] = path
+                    break
+
+        return output
+
+    def _make_datapipe(
+        self,
+        resource_dps: List[IterDataPipe],
+        *,
+        config: DatasetConfig,
+    ) -> IterDataPipe[Dict[str, Any]]:
+        archive_images, archive_targets = resource_dps
+
+        images_dp = Filter(archive_images, filter_fn=partial(self._filter_split_images, req_split=config.split))
+
+        images_dp = hint_sharding(images_dp)
+        images_dp = hint_shuffling(images_dp)
+
+        targets_dp = Grouper(
+            archive_targets,
+            group_key_fn=self._filter_group_targets,
+            buffer_size=INFINITE_BUFFER_SIZE,
+            group_size=4,
+        )
+
+        def key_fn(data: Tuple[Any, Any]) -> str:
+            data0 = data[0]
+            if isinstance(data0, tuple):
+                data0 = data0[0]
+            path = Path(data0)
+            # The pathes for images and targets are
+            # - leftImg8bit/val/frankfurt/frankfurt_000001_066574_leftImg8bit.png
+            # - gtFine/val/frankfurt/frankfurt_000001_066574_gtFine_polygons.json
+            # - gtFine/val/frankfurt/frankfurt_000001_066574_gtFine_labelIds.png
+            # we transform them into "frankfurt_000001_066574"
+            return "_".join(path.name.split("_", 3)[:3])
```

Comment:
Can't we use this also for the `Grouper`? This way we only have a single function that generates keys.